### PR TITLE
Replace Select component with a better one

### PR DESCRIPTION
**Pull request by iron3oxide**
### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/pynecone-io/pynecone/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/pynecone-io/pynecone/pulls ) for the desired changed?

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### New Feature Submission:

- [x] Does your submission pass the tests? 
- [x] Have you linted your code locally prior to submission?

Adds AdvancedSelect and AdvancedOption, which in tandem would enable stuff like searchable/ editable selects and MultiSelects. Wraps the `chakra-react-select` library.

Doesn't really work yet, I'm probably doing something wrong.

### CODE REVIEW

**Comment by iron3oxide** [comment]
This is the error message I'm usually getting:

```bash
poetry run pc run
───────────────────────── Starting Pynecone App ─────────────────────────
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/typer/main.py", line 532, in wrapper
    return callback(**use_params)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/pc.py", line 114, in run
    app = prerequisites.get_app()
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/utils/prerequisites.py", line 104, in get_app
    return __import__(module, fromlist=(constants.APP_VAR,))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/examples/example/example/example.py", line 42, in <module>
    app.add_page(index)
  File "/home/jason/PycharmProjects/pynecone/pynecone/app.py", line 270, in add_page
    raise e
  File "/home/jason/PycharmProjects/pynecone/pynecone/app.py", line 261, in add_page
    component = component if isinstance(component, Component) else component()
                                                                   ^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/examples/example/example/example.py", line 35, in index
    pc.aselect([3, 2]),
    ^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/components/forms/advanced_select.py", line 350, in create
    return super().create(*[], **props)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 368, in create
    return cls(children=children, **props)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 138, in __init__
    f"Invalid var passed for prop {key}, expected type {expected_type}, got value {value} of type {passed_type}."
  File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 291, in __repr__
    return format.json_dumps(self.render())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jason/PycharmProjects/pynecone/pynecone/utils/format.py", line 423, in json_dumps
    return json.dumps(obj, ensure_ascii=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/usr/lib64/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type set is not JSON serializable
```

**Comment by ElijahAhianyo** [comment]
> This is the error message I'm usually getting:
> 
> ```shell
> poetry run pc run
> ───────────────────────── Starting Pynecone App ─────────────────────────
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/typer/main.py", line 214, in __call__
>     return get_command(self)(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
>     return self.main(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1055, in main
>     rv = self.invoke(ctx)
>          ^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
>     return _process_result(sub_ctx.command.invoke(sub_ctx))
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/click/core.py", line 760, in invoke
>     return __callback(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/.cache/pypoetry/virtualenvs/pynecone-Yp1zm8LS-py3.11/lib/python3.11/site-packages/typer/main.py", line 532, in wrapper
>     return callback(**use_params)  # type: ignore
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/pc.py", line 114, in run
>     app = prerequisites.get_app()
>           ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/utils/prerequisites.py", line 104, in get_app
>     return __import__(module, fromlist=(constants.APP_VAR,))
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/examples/example/example/example.py", line 42, in <module>
>     app.add_page(index)
>   File "/home/jason/PycharmProjects/pynecone/pynecone/app.py", line 270, in add_page
>     raise e
>   File "/home/jason/PycharmProjects/pynecone/pynecone/app.py", line 261, in add_page
>     component = component if isinstance(component, Component) else component()
>                                                                    ^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/examples/example/example/example.py", line 35, in index
>     pc.aselect([3, 2]),
>     ^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/components/forms/advanced_select.py", line 350, in create
>     return super().create(*[], **props)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 368, in create
>     return cls(children=children, **props)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 138, in __init__
>     f"Invalid var passed for prop {key}, expected type {expected_type}, got value {value} of type {passed_type}."
>   File "/home/jason/PycharmProjects/pynecone/pynecone/components/component.py", line 291, in __repr__
>     return format.json_dumps(self.render())
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/jason/PycharmProjects/pynecone/pynecone/utils/format.py", line 423, in json_dumps
>     return json.dumps(obj, ensure_ascii=False)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/json/__init__.py", line 238, in dumps
>     **kw).encode(obj)
>           ^^^^^^^^^^^
>   File "/usr/lib64/python3.11/json/encoder.py", line 200, in encode
>     chunks = self.iterencode(o, _one_shot=True)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/json/encoder.py", line 258, in iterencode
>     return _iterencode(o, 0)
>            ^^^^^^^^^^^^^^^^^
>   File "/usr/lib64/python3.11/json/encoder.py", line 180, in default
>     raise TypeError(f'Object of type {o.__class__.__name__} '
> TypeError: Object of type set is not JSON serializable
> ```

There's no support for sets in json, so sets need to converted to lists in this case. May not be the cleanest solution but if you change the json dumps line in format.py to `json.dumps(obj, ensure_ascii=False, default=list)` that should work.

File: pynecone/components/forms/advanced_select.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+class AdvancedOption(Component):
+    """An option component for the chakra-react-select Select."""
+
+    library = "chakra-react-select"
+
+    tag = "Option"
+
+    # What is displayed to the user
+    label: Var[str]
```
Comment (picklelo) [inline pynecone/components/forms/advanced_select.py:18]: For non-components we don't need to use `Var`

File: pynecone/components/forms/advanced_select.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    #  "purple" | "pink" | "linkedin" | "facebook" | "messenger" |
+    #  "whatsapp" | "twitter" | "telegram"
+    # default: "gray"
+    color_scheme: Var[str]
+
+    # The icon of the option tag
+    icon: Var[str]
+
+    @classmethod
+    def create(cls, label: str, **props) -> Component:
```
Comment (picklelo) [inline pynecone/components/forms/advanced_select.py:41]: Then we can move this logic to the `__init__` function for non-components

File: pynecone/components/forms/advanced_select.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            options (List[Component | str | int | float | bool]): A list of values that can be of type
+        Component(AdvancedOption), str, int, float or bool.
+
+        Returns:
+            The `create` method is returning an instance of the `AdvancedSelect` class.
+        """
+        checked_options = []
+        for option in options:
+            if not isinstance(option, AdvancedOption):
+                checked_options.append(AdvancedOption.create(label=str(option)))
```
Comment (picklelo) [inline pynecone/components/forms/advanced_select.py:361]: Looks like the `options` prop expects a list of dicts which his the main problem

**Review by picklelo — COMMENTED** [review]
Great work! Wasn't sure how to add to this so I opened a new branch with some fixes that make it work: #1044

You can base your PR on that - once this is cleaned up I'm thinking we make this the default `pc.select` component since it's much better than the current one.

File: pynecone/components/forms/select.py [hunk 4]
```diff
@@ -1,53 +1,300 @@
...
+    color_scheme: Var[str]
 
-    # The default value of the select.
-    default_value: Var[str]
+    # This complex object includes all the compositional components
+    # that are used in react-select. If you wish to overwrite a component,
+    # pass in an object with the appropriate namespace.
+    # If you only wish to restyle a component,
+    # we recommend using the styles prop instead.
+    components: Var[dict[str, Component]]
```
Comment (picklelo) [inline pynecone/components/forms/select.py:124]: This needs to be `Dict[str, Component]` to support Python < 3.9

File: pynecone/components/forms/select.py [hunk 5]
```diff
@@ -1,53 +1,300 @@
...
+    name: Var[str]
+
+    # Allows control of whether the menu is opened when the Select is focused
+    open_menu_on_focus: Var[bool]
+
+    # Allows control of whether the menu is opened when the Select is clicked
+    open_menu_on_click: Var[bool]
+
+    # Array of options that populate the select menu
+    options: Var[list[Dict]]
```
Comment (picklelo) [inline pynecone/components/forms/select.py:218]: This needs to be `List[Dict]`

File: pynecone/components/forms/select.py [hunk 6]
```diff
@@ -59,51 +306,25 @@ def get_controlled_triggers(self) -> Dict[str, Var]:
...
-
-    value: Var[Any]
-
-    @classmethod
-    def create(cls, *children, **props) -> Component:
-        """Create a select option component.
-
-        By default, the value of the option is the text of the option.
+    def create(
+        cls, options: List[Option | str | int | float | bool], **props
```
Comment (picklelo) [inline pynecone/components/forms/select.py:310]: We need to use `List[Union[Option, str, int, float, bool]]`

**Review by picklelo — CHANGES_REQUESTED** [review]
Thanks for adding this, this is a huge improvement over the default one. Just a couple typing issues to get the build passing for Python < 3.9 and then we can merge this

**Comment by iron3oxide** [comment]
done

**Review by picklelo — APPROVED** [review]
awesome!

**Comment by masenf** [comment]
is there a doc update planned for this? seems like https://pynecone.io/docs/library/forms/select has bad information now.

**Pull request merged**